### ISSUE TITLE
updating to latest version (project_services fix) CRASHES terraform

### ISSUE THREAD

**Issue by ideasculptor** [terraform-google-modules/terraform-google-project-factory#305]
For projects that had active resources, re-applying with the latest module works fine, because it would error out on deleting at least one or two of the APIs when exchanging project_services for project_service in a loop.  Simply running terraform state rm on entries in question and then re-running the apply cleaned up the mess.

However, I had one project that was created with the older module, to which I had assigned no resources.  That one succeeded in the deletes, but then errored out once the compute api had been disabled.  I had to manually re-enable the compute api in the console, but when I attempted to re-apply the module, terraform now crashes immediately after refreshing state.

```
[terragrunt] 2019/10/31 11:33:13 Running command: terraform apply
data.terraform_remote_state.parent: Refreshing state...
data.terraform_remote_state.env: Refreshing state...
module.project.module.project-factory.random_id.random_project_id_suffix: Refreshing state... [id=oL8]
module.project.module.gsuite_group.data.google_organization.org[0]: Refreshing state...
module.project.module.project-factory.null_resource.shared_vpc_subnet_invalid_name[0]: Refreshing state... [id=aaaaaaaaa]
module.project.module.project-factory.null_resource.preconditions: Refreshing state... [id=bbbbbbbbbbbb]
module.project.module.project-factory.google_project.main: Refreshing state... [id=xxxxxxxx]
module.project.gsuite_group.group[0]: Refreshing state... [id=mmmmmmmmm]
gsuite_group_member.admin_group_member[0]: Refreshing state... [id=ccccccccc]
module.org-iam.google_organization_iam_member.organization_iam_additive[0]: Refreshing state... [id=rrrrrrrrrr/roles/resourcemanager.organizationViewer/group:xxxxxxxxxxxx-admin@domain.com]
module.folder-iam.google_folder_iam_member.folder_iam_additive[2]: Refreshing state... [id=folders/ffffffffffff/roles/iam.serviceAccountKeyAdmin/group:xxxxxxxxxx-admin@domain.com]
module.folder-iam.google_folder_iam_member.folder_iam_additive[6]: Refreshing state... [id=folders/gggggggg/roles/servicemanagement.admin/group:xxxxxxxx-admin@domain.com]
module.folder-iam.google_folder_iam_member.folder_iam_additive[7]: Refreshing state... [id=folders/495253307219/roles/serviceusage.serviceUsageAdmin/group:refarch-dev-svc-admin@ideasculptor.com]
module.folder-iam.google_folder_iam_member.folder_iam_additive[0]: Refreshing state... [id=folders/495253307219/roles/billing.projectManager/group:refarch-dev-svc-admin@ideasculptor.com]
module.folder-iam.google_folder_iam_member.folder_iam_additive[5]: Refreshing state... [id=folders/495253307219/roles/resourcemanager.projectCreator/group:refarch-dev-svc-admin@ideasculptor.com]
module.folder-iam.google_folder_iam_member.folder_iam_additive[3]: Refreshing state... [id=folders/495253307219/roles/owner/group:refarch-dev-svc-admin@ideasculptor.com]
module.folder-iam.google_folder_iam_member.folder_iam_additive[4]: Refreshing state... [id=folders/495253307219/roles/resourcemanager.folderAdmin/group:refarch-dev-svc-admin@ideasculptor.com]
module.folder-iam.google_folder_iam_member.folder_iam_additive[1]: Refreshing state... [id=folders/495253307219/roles/iam.serviceAccountAdmin/group:refarch-dev-svc-admin@ideasculptor.com]
module.project.module.project-factory.google_project_iam_member.gsuite_group_role[0]: Refreshing state... [id=refarch-dev-svc-a0bf/roles/editor/group:refarch-dev-svc-admin@ideasculptor.com]
module.project.module.project-factory.data.null_data_source.default_service_account: Refreshing state...
module.project.module.project-factory.module.project_services.google_project_service.project_services["iam.googleapis.com"]: Refreshing state... [id=refarch-dev-svc-a0bf/iam.googleapis.com]
module.project.module.project-factory.google_resource_manager_lien.lien[0]: Refreshing state... [id=p724103302051-ld27dca3f-5554-43e3-a9f4-305fff98a2f5]
module.project.module.project-factory.google_service_account.default_service_account: Refreshing state... [id=projects/refarch-dev-svc-a0bf/serviceAccounts/project-service-account@refarch-dev-svc-a0bf.iam.gserviceaccount.com]
module.project.module.project-factory.google_service_account_iam_member.service_account_grant_to_group[0]: Refreshing state... [id=projects/refarch-dev-svc-a0bf/serviceAccounts/project-service-account@refarch-dev-svc-a0bf.iam.gserviceaccount.com/roles/iam.serviceAccountUser/group:refarch-dev-svc-admin@ideasculptor.com]
module.project.gsuite_group_member.service_account_sa_group_member[0]: Refreshing state... [id=zzzzzzzzzzzzzzzzzz]
module.project.module.project-factory.google_project_iam_member.gke_host_agent[0]: Refreshing state... [id=xxxxxxxxxxxxx/roles/container.hostServiceAgentUser/serviceaccount:service-724103302051@container-engine-robot.iam.gserviceaccount.com]
module.project.module.project-factory.google_compute_shared_vpc_service_project.shared_vpc_attachment[0]: Refreshing state... [id=xxxxxxxxxxx/yyyyyyyyyyyy]
module.project.module.project-factory.google_project_usage_export_bucket.usage_report_export[0]: Refreshing state... [id=refarch-dev-svc-a0bf]
module.project.module.project-factory.null_resource.delete_default_compute_service_account[0]: Refreshing state... [id=7727890665502021964]
module.project.module.project-factory.google_project_iam_member.controlling_group_vpc_membership[1]: Refreshing state... [id=refarch-dev-e5ee/roles/compute.networkUser/serviceAccount:project-service-account@refarch-dev-svc-a0bf.iam.gserviceaccount.com]
module.project.module.project-factory.google_project_iam_member.controlling_group_vpc_membership[0]: Refreshing state... [id=refarch-dev-e5ee/roles/compute.networkUser/group:refarch-dev-svc-admin@ideasculptor.com]
module.project.module.project-factory.google_project_iam_member.controlling_group_vpc_membership[3]: Refreshing state... [id=refarch-dev-e5ee/roles/compute.networkUser/serviceaccount:service-724103302051@container-engine-robot.iam.gserviceaccount.com]
module.project.module.project-factory.google_project_iam_member.controlling_group_vpc_membership[2]: Refreshing state... [id=refarch-dev-e5ee/roles/compute.networkUser/serviceAccount:724103302051@cloudservices.gserviceaccount.com]

Error: rpc error: code = Unavailable desc = transport is closing



Error: rpc error: code = Unavailable desc = transport is closing



Error: rpc error: code = Unavailable desc = transport is closing



Error: rpc error: code = Unavailable desc = transport is closing



Error: rpc error: code = Unavailable desc = transport is closing



Error: rpc error: code = Unavailable desc = transport is closing



Error: rpc error: code = Unavailable desc = transport is closing


panic: runtime error: invalid memory address or nil pointer dereference
2019-10-31T11:33:19.136-0700 [DEBUG] plugin.terraform-provider-google_v2.18.1_x4: [signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x27c3919]
2019-10-31T11:33:19.136-0700 [DEBUG] plugin.terraform-provider-google_v2.18.1_x4: 
2019-10-31T11:33:19.136-0700 [DEBUG] plugin.terraform-provider-google_v2.18.1_x4: goroutine 144 [running]:
2019-10-31T11:33:19.136-0700 [DEBUG] plugin.terraform-provider-google_v2.18.1_x4: github.com/terraform-providers/terraform-provider-google/google.resourceProjectUsageBucketRead(0xc0005accb0, 0x2b3ea00, 0xc0003ca500, 0xc0005accb0, 0x0)
2019-10-31T11:33:19.136-0700 [DEBUG] plugin.terraform-provider-google_v2.18.1_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/terraform-providers/terraform-provider-google/google/resource_usage_export_bucket.go:60 +0x209
2019-10-31T11:33:19.136-0700 [DEBUG] plugin.terraform-provider-google_v2.18.1_x4: github.com/hashicorp/terraform-plugin-sdk/helper/schema.(*Resource).RefreshWithoutUpgrade(0xc000592380, 0xc0009f01e0, 0x2b3ea00, 0xc0003ca500, 0xc000a9e780, 0xc0009f01e0, 0x0)
2019-10-31T11:33:19.136-0700 [DEBUG] plugin.terraform-provider-google_v2.18.1_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.0.0/helper/schema/resource.go:455 +0x119
2019-10-31T11:33:19.136-0700 [DEBUG] plugin.terraform-provider-google_v2.18.1_x4: github.com/hashicorp/terraform-plugin-sdk/internal/helper/plugin.(*GRPCProviderServer).ReadResource(0xc00000e018, 0x32f0220, 0xc000a9e4e0, 0xc0009f00f0, 0xc00000e018, 0xc000a9e4e0, 0xc0007eba80)
2019-10-31T11:33:19.136-0700 [DEBUG] plugin.terraform-provider-google_v2.18.1_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.0.0/internal/helper/plugin/grpc_provider.go:525 +0x3d8
2019-10-31T11:33:19.137-0700 [DEBUG] plugin.terraform-provider-google_v2.18.1_x4: github.com/hashicorp/terraform-plugin-sdk/internal/tfplugin5._Provider_ReadResource_Handler(0x2e9f920, 0xc00000e018, 0x32f0220, 0xc000a9e4e0, 0xc0009104e0, 0x0, 0x32f0220, 0xc000a9e4e0, 0xc0005c6a00, 0xa0)
2019-10-31T11:33:19.137-0700 [DEBUG] plugin.terraform-provider-google_v2.18.1_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk@v1.0.0/internal/tfplugin5/tfplugin5.pb.go:3153 +0x217
2019-10-31T11:33:19.137-0700 [DEBUG] plugin.terraform-provider-google_v2.18.1_x4: google.golang.org/grpc.(*Server).processUnaryRPC(0xc0003cc160, 0x33206e0, 0xc000559080, 0xc0009bc100, 0xc000097c50, 0x40a5bb0, 0x0, 0x0, 0x0)
2019-10-31T11:33:19.137-0700 [DEBUG] plugin.terraform-provider-google_v2.18.1_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/google.golang.org/grpc@v1.23.0/server.go:995 +0x460
2019-10-31T11:33:19.137-0700 [DEBUG] plugin.terraform-provider-google_v2.18.1_x4: google.golang.org/grpc.(*Server).handleStream(0xc0003cc160, 0x33206e0, 0xc000559080, 0xc0009bc100, 0x0)
2019-10-31T11:33:19.137-0700 [DEBUG] plugin.terraform-provider-google_v2.18.1_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/google.golang.org/grpc@v1.23.0/server.go:1275 +0xd97
2019-10-31T11:33:19.137-0700 [DEBUG] plugin.terraform-provider-google_v2.18.1_x4: google.golang.org/grpc.(*Server).serveStreams.func1.1(0xc0000b2e80, 0xc0003cc160, 0x33206e0, 0xc000559080, 0xc0009bc100)
2019-10-31T11:33:19.137-0700 [DEBUG] plugin.terraform-provider-google_v2.18.1_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/google.golang.org/grpc@v1.23.0/server.go:710 +0xbb
2019-10-31T11:33:19.137-0700 [DEBUG] plugin.terraform-provider-google_v2.18.1_x4: created by google.golang.org/grpc.(*Server).serveStreams.func1
2019-10-31T11:33:19.137-0700 [DEBUG] plugin.terraform-provider-google_v2.18.1_x4: 	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/google.golang.org/grpc@v1.23.0/server.go:708 +0xa1
2019/10/31 11:33:19 [ERROR] module.project.module.project-factory: eval: *terraform.EvalRefresh, err: rpc error: code = Unavailable desc = transport is closing
2019/10/31 11:33:19 [ERROR] module.folder-iam: eval: *terraform.EvalRefresh, err: rpc error: code = Unavailable desc = transport is closing
2019/10/31 11:33:19 [ERROR] module.project.module.project-factory: eval: *terraform.EvalRefresh, err: rpc error: code = Unavailable desc = transport is closing
2019/10/31 11:33:19 [ERROR] module.project.module.project-factory: eval: *terraform.EvalSequence, err: rpc error: code = Unavailable desc = transport is closing
2019/10/31 11:33:19 [ERROR] module.project.module.project-factory: eval: *terraform.EvalSequence, err: rpc error: code = Unavailable desc = transport is closing
2019/10/31 11:33:19 [TRACE] [walkRefresh] Exiting eval tree: module.project.module.project-factory.google_project_usage_export_bucket.usage_report_export[0]
2019/10/31 11:33:19 [TRACE] vertex "module.project.module.project-factory.google_project_usage_export_bucket.usage_report_export[0]": visit complete
2019/10/31 11:33:19 [TRACE] vertex "module.project.module.project-factory.google_project_usage_export_bucket.usage_report_export": dynamic subgraph encountered errors
2019/10/31 11:33:19 [TRACE] vertex "module.project.module.project-factory.google_project_usage_export_bucket.usage_report_export": visit complete
2019-10-31T11:33:19.140-0700 [DEBUG] plugin: plugin process exited: path="/Users/sgendler/src/gcloud-reference-arch/Reference Infrastructure/dev/service-project/.terragrunt-cache/bM5DHXZHR8tYV2VLTKMTZaYeslQ/Q1O00TrLpfRM585ghG_iSMYWaSM/service-project/environment/.terraform/plugins/darwin_amd64/terraform-provider-google_v2.18.1_x4" pid=96979 error="exit status 2"
2019/10/31 11:33:19 [TRACE] [walkRefresh] Exiting eval tree: module.project.module.project-factory.google_project_iam_member.controlling_group_vpc_membership[2]
2019/10/31 11:33:19 [TRACE] vertex "module.project.module.project-factory.google_project_iam_member.controlling_group_vpc_membership[2]": visit complete
2019/10/31 11:33:19 [ERROR] module.folder-iam: eval: *terraform.EvalSequence, err: rpc error: code = Unavailable desc = transport is closing
2019/10/31 11:33:19 [TRACE] [walkRefresh] Exiting eval tree: module.folder-iam.google_folder_iam_member.folder_iam_additive[1]
2019/10/31 11:33:19 [TRACE] vertex "module.folder-iam.google_folder_iam_member.folder_iam_additive[1]": visit complete
2019/10/31 11:33:19 [ERROR] module.project.module.project-factory: eval: *terraform.EvalRefresh, err: rpc error: code = Unavailable desc = transport is closing
2019/10/31 11:33:19 [TRACE] dag/walk: upstream of "root" errored, so skipping
2019/10/31 11:33:19 [ERROR] module.project.module.project-factory: eval: *terraform.EvalRefresh, err: rpc error: code = Unavailable desc = transport is closing
2019/10/31 11:33:19 [TRACE] vertex "module.folder-iam.google_folder_iam_member.folder_iam_additive": dynamic subgraph encountered errors
2019/10/31 11:33:19 [ERROR] module.project.module.project-factory: eval: *terraform.EvalSequence, err: rpc error: code = Unavailable desc = transport is closing
2019/10/31 11:33:19 [TRACE] vertex "module.folder-iam.google_folder_iam_member.folder_iam_additive": visit complete
2019/10/31 11:33:19 [TRACE] [walkRefresh] Exiting eval tree: module.project.module.project-factory.google_project_iam_member.controlling_group_vpc_membership[0]
2019/10/31 11:33:19 [TRACE] vertex "module.project.module.project-factory.google_project_iam_member.controlling_group_vpc_membership[0]": visit complete
2019/10/31 11:33:19 [ERROR] module.project.module.project-factory: eval: *terraform.EvalRefresh, err: rpc error: code = Unavailable desc = transport is closing
2019/10/31 11:33:19 [ERROR] module.project.module.project-factory: eval: *terraform.EvalSequence, err: rpc error: code = Unavailable desc = transport is closing
2019/10/31 11:33:19 [TRACE] [walkRefresh] Exiting eval tree: module.project.module.project-factory.google_project_iam_member.controlling_group_vpc_membership[1]
2019/10/31 11:33:19 [ERROR] module.project.module.project-factory: eval: *terraform.EvalSequence, err: rpc error: code = Unavailable desc = transport is closing
2019/10/31 11:33:19 [TRACE] vertex "module.project.module.project-factory.google_project_iam_member.controlling_group_vpc_membership[1]": visit complete
2019/10/31 11:33:19 [TRACE] [walkRefresh] Exiting eval tree: module.project.module.project-factory.google_project_iam_member.controlling_group_vpc_membership[3]
2019/10/31 11:33:19 [TRACE] vertex "module.project.module.project-factory.google_project_iam_member.controlling_group_vpc_membership[3]": visit complete
2019/10/31 11:33:19 [ERROR] module.project.module.project-factory: eval: *terraform.EvalRefresh, err: rpc error: code = Unavailable desc = transport is closing
2019/10/31 11:33:19 [TRACE] dag/walk: upstream of "root" errored, so skipping
2019/10/31 11:33:19 [ERROR] module.project.module.project-factory: eval: *terraform.EvalSequence, err: rpc error: code = Unavailable desc = transport is closing
2019/10/31 11:33:19 [TRACE] vertex "module.project.module.project-factory.google_project_iam_member.controlling_group_vpc_membership": dynamic subgraph encountered errors
2019/10/31 11:33:19 [TRACE] [walkRefresh] Exiting eval tree: module.project.module.project-factory.google_compute_shared_vpc_service_project.shared_vpc_attachment[0]
2019/10/31 11:33:19 [TRACE] vertex "module.project.module.project-factory.google_project_iam_member.controlling_group_vpc_membership": visit complete
2019/10/31 11:33:19 [TRACE] vertex "module.project.module.project-factory.google_compute_shared_vpc_service_project.shared_vpc_attachment[0]": visit complete
2019/10/31 11:33:19 [TRACE] vertex "module.project.module.project-factory.google_compute_shared_vpc_service_project.shared_vpc_attachment": dynamic subgraph encountered errors
2019/10/31 11:33:19 [TRACE] vertex "module.project.module.project-factory.google_compute_shared_vpc_service_project.shared_vpc_attachment": visit complete
2019/10/31 11:33:19 [TRACE] dag/walk: upstream of "provider.google (close)" errored, so skipping
2019/10/31 11:33:19 [TRACE] dag/walk: upstream of "root" errored, so skipping
2019-10-31T11:33:19.278-0700 [DEBUG] plugin: plugin exited



!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

Terraform crashed! This is always indicative of a bug within Terraform.
A crash log has been placed at "crash.log" relative to your current
working directory. It would be immensely helpful if you could please
report the crash with Terraform[1] so that we can fix this.

When reporting bugs, please include your terraform version. That
information is available on the first line of crash.log. You can also
get it by running 'terraform --version' on the command line.

[1]: https://github.com/hashicorp/terraform/issues

!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!
[terragrunt] 2019/10/31 11:33:19 Hit multiple errors:
exit status 1
```

```
$ terraform --version
Terraform v0.12.12
```

**Comment by ideasculptor** [comment]
The really fun part is that even running `terragrunt destroy` doesn't fix it - it crashes, too.

**Comment by morgante** [comment]
A Terraform crash is likely beyond the scope of the module - that should "never" happen from resources. Can you file an issue on the provider? https://github.com/terraform-providers/terraform-provider-google/issues

As a fix, you might want to try manually activating *only* the APIs with this command:

- `terraform apply -refresh=false -target=module.project.module.project-factory.module. project_services`

**Comment by ideasculptor** [comment]
I tried that.  I tried manually enabling all of the APIs it used to have.  I made sure no trace of the old project_services resource was still in the state file - terraform crashed, regardless - on both apply and delete.  I eventually had to manually delete the project, delete the state, and apply from scratch in order to get anything to work.

I suspect that the module needs a migration strategy which creates the new project_service resources in the module and does a state rm of the old ones.
